packages/token-search-discovery-controller/src/token-search-discovery-controller.ts

```diff
@@ -0,0 +1,122 @@
+import type { RestrictedControllerMessenger } from '@metamask/base-controller';
+import { BaseController } from '@metamask/base-controller';
+import type { TokenSearchResponseItem } from './types';
+
+export const TOKENSEARCH_EVENTS = {
+  SEARCH_COMPLETED: 'SEARCH_COMPLETED',
+} as const;
+
+const name = 'TokenSearchDiscoveryController';
+
+type Events = {
+  type: (typeof TOKENSEARCH_EVENTS)[keyof typeof TOKENSEARCH_EVENTS];
+  payload: [TokenSearchResponseItem[]];
+};
+
+export type TokenSearchDiscoveryControllerMessenger =
+  RestrictedControllerMessenger<
+    typeof name,
+    never,
+    Events,
+    never,
+    (typeof TOKENSEARCH_EVENTS)[keyof typeof TOKENSEARCH_EVENTS]
+  >;
+
+export type TokenSearchDiscoveryState = {
+  recentSearches: TokenSearchResponseItem[];
+  lastSearchTimestamp: number | null;
+};
+
+const defaultState: TokenSearchDiscoveryState = {
+  recentSearches: [],
+  lastSearchTimestamp: null,
+};
```

Comment:
Default state should be returned via a function and not stored in a variable. This makes it handy to use in tests (and sometimes default state needs to be dynamic):

``` suggestion
/**
 * Constructs the default {@link TokenSearchDiscoveryController} state. This allows
 * consumers to provide a partial state object when initializing the controller
 * and also helps in constructing complete state objects for this controller in
 * tests.
 *
 * @returns The default {@link TokenSearchDiscoveryController} state.
 */
export function getDefaultTokenSearchDiscoveryControllerState(): TokenSearchDiscoveryControllerState {
  return {
    recentSearches: [],
    lastSearchTimestamp: null,
  };
}